[PM-18044] Ensure all calls to receive should get all messages broadcast

coroiu opened the pull request:
## 🎟️ Tracking

<!-- Paste the link to the Jira or GitHub issue or otherwise describe / point to where this change is coming from. -->

## 📔 Objective

<!-- Describe what the purpose of this PR is, for example what bug you're fixing or new feature you're adding. -->

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation
  team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

coroiu committed:
feat: implement communication backend with tokio channel

coroiu committed:
feat: implement ability to subscribe

coroiu committed:
fix: lint

dani-garcia commented on crates/bitwarden-ipc/src/traits/communication_backend.rs:
```diff
@@ -20,7 +20,27 @@ pub trait CommunicationBackend {
+    ///
+    /// Use this function if you are expecting to receive a message as a response to a message you
+    /// sent. Subscribing will start buffering messages and help you avoid missing messages.
+    fn subscribe(&self) -> impl CommunicationBackendReceiver<ReceiveError = Self::ReceiveError>;
```
As far as I understand it, the invariant that all the subscribers will receive all the same messages is something that the trait implementations need to enforce right? If so it might be a good idea to mention it in here. Otherwise someome might think that it's the `IpcClient` that handles that for you, leading to inconsistent implementations.

The same also probably applies to `CommunicationBackendReceiver::receive`

coroiu replied:
Good point, I've updated the doc, thoughts?

coroiu committed:
docs: clarify implementation expectations

dani-garcia commented on crates/bitwarden-ipc/src/wasm/communication_backend.rs:
```diff
@@ -1,53 +1,102 @@
+
+    async fn receive(&self) -> Result<IncomingMessage, Self::ReceiveError> {
+        Ok(self
+            .resubscribe()
```
This doesn't feel entirely correct, the docs for `resubscribe` say:

    /// Re-subscribes to the channel starting from the current tail element.
    ///
    /// This [`Receiver`] handle will receive a clone of all values sent
    /// **after** it has resubscribed. This will not include elements that are
    /// in the queue of the current receiver. Consider the following example.


Which mean each call to `receive` will never get any buffered values, and only receive any values after `receive` has been called.

Removing the resubscribe causes errors as receive requires &mut, so we might need to do something like:

```
impl CommunicationBackend for JsCommunicationBackend {
    ...

    fn subscribe(&self) -> impl CommunicationBackendReceiver<ReceiveError = Self::ReceiveError> {
        Mutex::new(self.receive_rx.resubscribe())
    }
}

impl CommunicationBackendReceiver for Mutex<tokio::sync::broadcast::Receiver<IncomingMessage>> {
    type ReceiveError = JsValue;

    async fn receive(&self) -> Result<IncomingMessage, Self::ReceiveError> {
        Ok(self
            .lock()
            .await
            .recv()
            .await
            .map_err(|e| ChannelError(e.to_string()))?)
    }
}
```

coroiu replied:
Ouf good point, I didn't realize this changed by moving the buffering from the platform to Rust, I'll address this asap 👍

dani-garcia replied:
This backend receiver is still resubscribing on receive, which means it won't receive any previous messages, probably needs to be updated to be similar to `TestCommunicationBackendReceiver`

coroiu replied:
Yeah, noticed this in a follow-up PR at the same time as you wrote this comment 🥲 Will fix

coroiu committed:
feat: refactor communication backend trait

coroiu committed:
refactor: remove general `receive()` function

`receive` has been removed in favor of requiring consumers to `subscribe`. This improves thread-handling

coroiu committed:
feat: refactor IpcClient to use subscriptions

coroiu committed:
feat: improve client API and fix backend testing

- Client API is now much more fluent
- The communication backend testing utilty did not properly emulated
the real behavior of a subscription based channel where messages
before a subscription occured would be discarded.

coroiu committed:
chore: remove debug prints

coroiu committed:
chore: clean up commented code

coroiu committed:
feat: add topic to subcription

This will avoid weirdness where it's not obvious that calling `receive(Some(topic))` will actually discard all messages not matching that topic. This way it'll be completely separate subscriptions behaving more intuitively

coroiu committed:
feat: split subscriptions into typed and untyped

coroiu committed:
fix: remove stale doc comment

coroiu committed:
fix: lint

dani-garcia commented on crates/bitwarden-ipc/src/ipc_client.rs:
```diff
@@ -72,26 +140,75 @@ where
+            <Com::Receiver as CommunicationBackendReceiver>::ReceiveError,
+        >,
+    > {
+        self.client.receive_typed(&self.receiver, timeout).await
```
I wonder if we should inline the `receive_typed` method from `IpcClient` here, it's only used here and it's simple enough:
```
    pub async fn receive(
        &self,
        timeout: Option<Duration>,
    ) -> Result<
        TypedIncomingMessage<Payload>,
        TypedReceiveError<
            <Payload as TryFrom<Vec<u8>>>::Error,
            Crypto::ReceiveError,
            <Com::Receiver as CommunicationBackendReceiver>::ReceiveError,
        >,
    > {
        let topic = Some(Payload::name());
        let received = self.client.receive(&self.receiver, &topic, timeout).await?;
        received.try_into().map_err(TypedReceiveError::Typing)
    }
```

coroiu replied:
I was thinking that the req/res would build upon this and so it made sense to keep it separate, but I ended up inlining it in a follow up PR so maybe we can just merge this as is 🤷‍♂️

dani-garcia commented on crates/bitwarden-ipc/src/traits/communication_backend.rs:
```diff
@@ -65,19 +111,33 @@ pub mod tests {
+                .await
+                .recv()
+                .await
+                .expect("Failed to receive incoming message"))
```
We should grab the rwlock only once to avoid issues with other requests getting in between. Also use `.is_empty()` instead of `.len() == 0`

```suggestion
            let mut guard = self.0.write().await;

            if guard.is_empty() {
                return Err(TestCommunicationBackendReceiveError::NoQueuedMessages);
            }

            Ok(guard
                .recv()
                .await
                .expect("Failed to receive incoming message"))
        }
```

coroiu committed:
fix: race condition issue

coroiu committed:
fix: CommunicationBackendReceiver resubscribing
